[RDY] Fix crash on invalid modelines

florolf opened the pull request:
fixes #5449

florolf committed:
charset: add getdigits_safe

Add a version of getdigits which does not crash on overflows, but
reports them to the caller and reimplement getdigits using that
function.

florolf committed:
buffer: use safe modeline version handling

A file containing the string "vim" followed by a very large number in a
location that can contain modelines will trigger an overflow in
getdigits() which is called by chk_modeline() when trying to parse the
version number.

Replace getdigits() by getdigits_safe() to avoid this issue.

florolf commented on src/nvim/charset.c:
```diff
@@ -1740,11 +1759,8 @@ char_u* skiptowhite_esc(char_u *p) {
 {
-  errno = 0;
-  intmax_t number = strtoimax((char *)*pp, (char **)pp, 10);
-  if (number == INTMAX_MAX || number == INTMAX_MIN) {
```
The check for `INTMAX_MAX`/`INTMAX_MIN` is a little redundant if we reset `errno` before. On the other hand, a strict reading of `strtoimax(3p)` would theoretically allow an implementation to set `errno` to `ERANGE` even if an overflow *didn't* occur (returning a value other than the extremal ones). I'll re-add the check.

jamessan commented:
There are some lint failures:

```
src/nvim/buffer.c:4536:  if should always use braces  [readability/braces] [5]

src/nvim/buffer.c:4539:  if should always use braces  [readability/braces] [5]

src/nvim/charset.c:1745:  Missing space before asterisk in )*  [whitespace/operators] [2]

src/nvim/charset.c:1747:  Boolean operator should be placed on the same line as the start of its right operand  [whitespace/operators] [4]
```

Could you also add a test?

jamessan commented on src/nvim/charset.c:
```diff
@@ -1740,11 +1760,8 @@ char_u* skiptowhite_esc(char_u *p) {
-    assert(errno != ERANGE);
-  }
+  intmax_t number;
+  assert(getdigits_safe(pp, &number) == 0);
```
Yay, QB saw that too (for those days when oni-link's scrutiny isn't present).

florolf replied:
Good catch, thanks.

I'm currently traveling and I'll address all the comments once I'm back.

florolf committed:
buffer: fix unrelated lint failures

While this PR didn't directly touch this line, the linter started
complaining about it as a result. Fix it nonetheless.

florolf committed:
add functional test for issue #5449

vim used to crash when the version number given in a mode line was too
large. Add a regression test for that scenario.

justinmk commented on src/nvim/charset.c:
```diff
@@ -1740,11 +1760,15 @@ char_u* skiptowhite_esc(char_u *p) {
+
+  // prevent "unused variable" warning if we are not doing a debug build
+  (void) ret;
+  assert(ret == 0);
```
could do instead

```c
if (ret != 0) {
    abort();
}
```

justinmk replied:
It's better than a silent bug, usually

justinmk commented on src/nvim/charset.c:
```diff
@@ -1732,6 +1732,26 @@ char_u* skiptowhite_esc(char_u *p) {
+    return -1;
+  }
+
+  return 0;
```
Returning OK/FAIL or `true`/`false` would be more idiomatic for the project I believe. Also seems to be what other functions in charset.c are doing.

florolf replied:
The other functions which return `true`/`false` are predicates, as far as I can tell. Using booleans for reporting failures always seemed a little "un-C-ish" to me, but I don't know the neovim codebase well enough. If you think otherwise, I'll change it.

justinmk replied:
OK/FAIL/NOTDONE (from `vim.h`) is not a boolean, any objection to that? 

Most important is to try to keep the pattern the same. OK=1, if we return 0 in some places to mean "ok", that can be unsettling.

florolf replied:
Okay, sounds sensible.

jamessan commented on src/nvim/buffer.c:
```diff
@@ -4528,16 +4528,20 @@ chk_modeline (
         else
           e = s + 3;
-        vers = getdigits_int(&e);
+        if (getdigits_safe(&e, &vers) < 0) {
```
This is a change in behavior from Vim.  This causes nvim to skip the modeline entirely, where as Vim ends up getting `vers = -1` since it downcasts the long to an int and then uses that in the comparison on line 4539.  The comparison succeeds and the modeline is set.

Now, one can question whether that's the correct behavior for Vim (and I would suggest bringing it up there), but that still leaves open how nvim should behave.  I'm in favor of ignoring the modeline, since I think it's better to be overcautious with modelines, but we should also be aware of and intentional about places where we deviate in behavior from Vim.

justinmk replied:
Also regarding the specific code here, it should check `!= OK`. I fixed this during the merge.

justinmk commented:
Merged. Thanks @florolf
